<a href="https://colab.research.google.com/github/Kenza-code/Analyse-des-sentiments-NLP---Avis-sur-Amazon-/blob/main/Analyse_des_sentiments_Amazon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Outils de programmation IA
Réalisé par : Kenza IDIR

# INTRODUCTION

Les avis en ligne ont un impact considérable sur les décisions d'achat des consommateurs. Cette étude vise à analyser les sentiments exprimés dans les avis clients d'Amazon afin de mieux comprendre les attentes des consommateurs et d'identifier les facteurs clés influençant leur satisfaction. En m'appuyant sur un ensemble de données comprenant  d'avis sur divers produits, Je cherche à analyser les sentiments des clients sur leurs avis vis à vis des produits d'Amazon

Les objectifs principaux de ce projet sont :

- Explorer la distribution des sentiments positifs, négatifs et neutres dans les avis.
- Développer un modèle de classification de sentiments afin d'automatiser l'analyse de nouveaux avis.

En utilisant cet ensemble de données que j'ai téléchargé sur kaggle https://www.kaggle.com/datasets/miriamodeyianypeter/sentiment-analysis-amazon-product-reviews/data , j'ai appliqué des techniques de traitement du langage naturel pour extraire des informations pertinentes et quantifier les sentiments exprimés. Les résultats de cette étude permettront de mieux comprendre les attentes des consommateurs et d'orienter les stratégies marketing et de développement de produits.



# **CHARGEMENT DES DONNEES**

In [ ]:
#Importation des bibliothèques nécessaires
import pandas as pd

In [ ]:
amazon = pd.read_csv("/content/Amazon-Product-Reviews - Amazon Product Review (1).csv")

In [ ]:
amazon.head()

marketplace  customer_id       review_id  product_id  product_parent  \
0          US     11555559  R1QXC7AHHJBQ3O  B00IKPX4GY         2693241   
1          US     31469372  R175VSRV6ZETOP  B00IKPYKWG         2693241   
2          US     26843895  R2HRFF78MWGY19  B00IKPW0UA         2693241   
3          US     19844868   R8Q39WPKYVSTX  B00LCHSHMS         2693241   
4          US      1189852  R3RL4C8YP2ZCJL  B00IKPZ5V6         2693241   

                           product_title product_category  star_rating  \
0  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            5   
1  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            3   
2  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            5   
3  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            4   
4  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            5   

   helpful_votes  total_votes vine verified_purchase  \
0              0            0    N                 Y   
1              0            0    N                 N   
2              0            0    N                 Y   
3              0            0    N                 N   
4              0            0    N                 Y   

                                     review_headline  \
0                                         Five Stars   
1  Lots of ads Slow processing speed Occasionally...   
2                            Well thought out device   
3  Not all apps/games we were looking forward to ...   
4                                         Five Stars   

                                         review_body review_date  sentiment  
0                                      Great love it  2015-08-31          1  
1  Lots of ads<br />Slow processing speed<br />Oc...  2015-08-31          0  
2  Excellent unit.  The versatility of this table...  2015-08-31          1  
3  I bought this on Amazon Prime so I ended up bu...  2015-08-31          1  
4  All Amazon products continue to meet my expect...  2015-08-31          1

In [ ]:
#convertir le jeu de données en un dataframe
amazon = pd.DataFrame(amazon)
#Aperçu des données
amazon.head()

marketplace  customer_id       review_id  product_id  product_parent  \
0          US     11555559  R1QXC7AHHJBQ3O  B00IKPX4GY         2693241   
1          US     31469372  R175VSRV6ZETOP  B00IKPYKWG         2693241   
2          US     26843895  R2HRFF78MWGY19  B00IKPW0UA         2693241   
3          US     19844868   R8Q39WPKYVSTX  B00LCHSHMS         2693241   
4          US      1189852  R3RL4C8YP2ZCJL  B00IKPZ5V6         2693241   

                           product_title product_category  star_rating  \
0  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            5   
1  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            3   
2  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            5   
3  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            4   
4  Fire HD 7, 7" HD Display, Wi-Fi, 8 GB               PC            5   

   helpful_votes  total_votes vine verified_purchase  \
0              0            0    N                 Y   
1              0            0    N                 N   
2              0            0    N                 Y   
3              0            0    N                 N   
4              0            0    N                 Y   

                                     review_headline  \
0                                         Five Stars   
1  Lots of ads Slow processing speed Occasionally...   
2                            Well thought out device   
3  Not all apps/games we were looking forward to ...   
4                                         Five Stars   

                                         review_body review_date  sentiment  
0                                      Great love it  2015-08-31          1  
1  Lots of ads<br />Slow processing speed<br />Oc...  2015-08-31          0  
2  Excellent unit.  The versatility of this table...  2015-08-31          1  
3  I bought this on Amazon Prime so I ended up bu...  2015-08-31          1  
4  All Amazon products continue to meet my expect...  2015-08-31          1

In [ ]:
amazon.shape

(30846, 16)

In [ ]:
amazon.columns

Index(['marketplace', 'customer_id', 'review_id', 'product_id',
       'product_parent', 'product_title', 'product_category', 'star_rating',
       'helpful_votes', 'total_votes', 'vine', 'verified_purchase',
       'review_headline', 'review_body', 'review_date', 'sentiment'],
      dtype='object')

In [ ]:
#Vérification des colonnes et de leurs contenu
print(amazon.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30846 entries, 0 to 30845
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   marketplace        30846 non-null  object
 1   customer_id        30846 non-null  int64 
 2   review_id          30846 non-null  object
 3   product_id         30846 non-null  object
 4   product_parent     30846 non-null  int64 
 5   product_title      30846 non-null  object
 6   product_category   30846 non-null  object
 7   star_rating        30846 non-null  int64 
 8   helpful_votes      30846 non-null  int64 
 9   total_votes        30846 non-null  int64 
 10  vine               30846 non-null  object
 11  verified_purchase  30846 non-null  object
 12  review_headline    30844 non-null  object
 13  review_body        30842 non-null  object
 14  review_date        30846 non-null  object
 15  sentiment          30846 non-null  int64 
dtypes: int64(6), object(10)
memory usage: 3.

# Préparation du jeu de données

#Elimination des colonnes

Dans ce jeu de données, il y a des colonnes qui sont susceptibles de ne pas être utiles pour l'analyse des sentiments, et donc elles peuvent être potentiellement supprimées :

Voici les colonnes que j'ai envisagé de supprimer :
- marketplace, customer_id, review_id, product_id, product_parent : ces colonnes contiennent des identifiants uniques et peuvent ne pas être directement pertinentes pour l'analyse des sentiments.
- product_title, product_category, vine, verified_purchase : bien que ces colonnes puissent fournir un contexte, elles ne sont pas forcément nécessaires à l'analyse basique des sentiments. Elles peuvent toutefois être utiles pour des analyses plus avancées, comme l'identification des tendances de sentiment spécifiques à un produit.
review_date : cette colonne peut être utile pour l'analyse des séries chronologiques des tendances de sentiment, mais peut ne pas être essentielle pour la classification de base des sentiments.
- star_rating : cela peut être utilisé comme une note numérique, fournissant un signal de sentiment direct mais comme je vais utiliser la colonne qui contient le texte réel de l'avis, donc je n'aurai pas besoin de la garder.
- review_headline : cela peut offrir un contexte supplémentaire et des indices de sentiment, en particulier pour les critiques plus courtes mais vu qu'il y a la colonne review_body qui contient le corps principal de l'avis, où le client exprime son opinion de manière détaillée et qui décrit les expériences du consommateur avec le produit, ses points forts et ses points faibles, et justifier sa note donc je n'aurai pas besoin de la garder.
- Helpful_votes et Total_votes : Elles ne sont pas des colonnes directement liés au sentiment, mais elles peuvent fournir des informations sur la qualité et l'influence de l'avis.

Et voici les colonnes que j'ai gardé :
- review_body : cette colonne contient le texte réel de l'avis, qui constitue la principale source d'informations pour l'analyse des sentiments.
- sentiment : cette colonne fournit une évaluation pré-établie du sentiment exprimé dans chaque avis. Cependant, il est important d'aborder cette colonne.


In [ ]:
df = amazon[["review_body", "sentiment"]]
df.head()

review_body  sentiment
0                                      Great love it          1
1  Lots of ads<br />Slow processing speed<br />Oc...          0
2  Excellent unit.  The versatility of this table...          1
3  I bought this on Amazon Prime so I ended up bu...          1
4  All Amazon products continue to meet my expect...          1

In [ ]:
#Vérification des valeurs manquantes
print(df.isnull().sum())

review_body    4
sentiment      0
dtype: int64


In [ ]:
#Suppression des données manquantes
df = df.dropna()
#Vérification
df.isnull().sum()

review_body    0
sentiment      0
dtype: int64

In [ ]:
#Vérification de la présence des lignes dupliquées
print(df.duplicated().sum())

3053


In [ ]:
#Supprimer les lignes dupliquées
df = df.drop_duplicates()
print(df.duplicated().sum())

0


In [ ]:
df.shape

(30842, 2)

# Prétraitement des données

Dans cette partie, je nettoierai le texte pour supprimer les éléments inutiles comme les ponctuations, les liens, ou les stopwords (car Ils sont considérés comme du "bruit" dans les analyses de texte, ils ne contribuent pas significativement à comprendre le sens d'une phrase)

In [ ]:
import re
import nltk
from nltk.corpus import stopwords
from sklearn.model_selection import train_test_split

#Je télécharge les stopwords
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

def clean_text(text):
  """
  Cette fonction nettoie le texte en supprimant les ponctuations,
  les caractères spéciaux, les liens et les stopwords.
  """
  text = re.sub(r'http\S+', '', text) #Suppression des urls
  text = re.sub(r'[^a-zA-Z]', ' ', text) #Suppression des carctères spéciaux
  text = text.lower() #Convertir en miniscule
  text = ' '.join([word for word in text.split() if word not in stop_words]) #Supprimer les stopwords
  return text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
#Je nettoie les avis
df["cleaned_review"] = df["review_body"].apply(clean_text)

#Division des données

Dans cette partie, je divise les données en un ensemble d'entraînement et de test

In [ ]:
X = df["cleaned_review"]
y = df["sentiment"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Vectorisation des données

Dans cette section, je convertis les textes en vecteurs numériques à l'aide de TF-IDF

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
# Initialiser le vectoriseur TF-IDF
tfidf = TfidfVectorizer(max_features=5000)  # Limitation à 5000 mots les plus fréquents

# Transformer les données
X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

print(f"Dimensions après vectorisation : {X_train_tfidf.shape}, {X_test_tfidf.shape}")

Dimensions après vectorisation : (24673, 5000), (6169, 5000)


#Modélisation

# Modèles d'apprentissage Automatique

Dans cette partie, je vais entraîner et comparer plusieurs modèles d'apprentissage automatique tout en expliquant mon choix des modèles d'apprentissage automatique que je développerai

Pourquoi j'ai choisi de comparer ces modèles pour choisir le modèle le plus performant dans mon cas d'études ?

En effet, j'ai choisi d'implémenter des modèles particulièrement adaptés pour ce genre de tâche (Analyse des sentiments des consommateurs des produits Amazon)

1. Régression Logistique (Logistic Regression)
La régression logistique est un modèle linéaire simple et efficace pour les problèmes de classification binaire. Dans le cas d'une analyse de sentiments, il peut prédire si un avis est "positif" ou "négatif".
Dans le jeu de données que j'ai choisi, il existe des relations linéaires entre les caractéristiques (plus précisément, les mots dans un avis).Ce modèle est bien adapté lorsque les données sont linéairement séparables.  
La régression logistique peut être utilisée en combinaison avec des techniques de vectorisation de texte (comme TF-IDF) pour transformer les avis en caractéristiques numériques.


2. Machine à Vecteurs de Support (SVM)
Le SVM est puissant pour des tâches de classification binaire et fonctionne bien dans des espaces de caractéristiques de haute dimension, comme les données textuelles (ce qui est le cas dans le jeu de données choisi)
SVM cherche à maximiser la frontière entre les classes (par exemple, sentiment positif vs négatif) et peut être particulièrement efficace lorsqu'il y a une séparation claire dans les données, même dans des espaces non linéaires grâce à l'utilisation du noyau.
Il est robuste aux sur-ajustements (overfitting).

3. Forêts Aléatoires (Random Forests)
Random Forest est un ensemble d'arbres de décision et peut être utilisé pour des tâches de classification en exploitant des sous-ensembles aléatoires de données pour entraîner plusieurs arbres. Cela rend le modèle robuste et moins susceptible au sur-ajustement.
Pour l'analyse de sentiments, il peut gérer efficacement les relations complexes entre les caractéristiques extraites des avis.

3. K-Plus Proches Voisins (KNN)
KNN peut être utile lorsque la similarité entre les avis (par exemple, les critiques ayant un sentiment similaire) est importante pour la classification.
Cependant, KNN peut devenir lent à traiter de grands ensembles de données, car il nécessite de calculer la distance entre tous les points lors de la prédiction.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score, accuracy_score

In [ ]:
# Initialisation des modèles
models = {
    "Logistic Regression": LogisticRegression(),
    "SVM": SVC(kernel='linear'),
    "Random Forest": RandomForestClassifier(n_estimators=100),
    "KNN": KNeighborsClassifier(n_neighbors=3)
}

# Entraînement et évaluation
results = {}
for model_name, model in models.items():
    print(f"Entraînement de {model_name}")
    model.fit(X_train_tfidf, y_train)
    y_pred = model.predict(X_test_tfidf)
    results[model_name] = {
        "Accuracy": accuracy_score(y_test, y_pred),
        "F1-Score": f1_score(y_test, y_pred, average='weighted')
    }

# Affichage des résultats
print("Comparaison des modèles :")
for model, metrics in results.items():
    print(f"{model} -> Accuracy: {metrics['Accuracy']:.4f}, F1-Score: {metrics['F1-Score']:.4f}")

Entraînement de Logistic Regression
Entraînement de SVM
Entraînement de Random Forest
Entraînement de KNN
Comparaison des modèles :
Logistic Regression -> Accuracy: 0.9100, F1-Score: 0.9044
SVM -> Accuracy: 0.9108, F1-Score: 0.9072
Random Forest -> Accuracy: 0.8933, F1-Score: 0.8809
KNN -> Accuracy: 0.8479, F1-Score: 0.7966


# Réseaux de neurones

Dans cette partie, je vais entraîner et comparer plusieurs réseaux de neurones tout en expliquant pourqui j'ai choisi de développer ces réseaux de neurones et les comparer avec les modèles d'apprentissage automatique précédents

Pourquoi j'ai choisi de comparer ces réseaux pour choisir le modèle le plus performant dans mon cas d'études ?

En effet, j'ai choisi d'implémenter des réseaux de neurones particulièrement adaptés pour ce genre de tâche (Analyse des sentiments des consommateurs des produits Amazon)

1. Réseau de Neurones Dense (DNN)
Les réseaux de neurones sont capables d'apprendre des relations complexes et non linéaires dans les données. Ils sont souvent utilisés pour des tâches de classification plus complexes.
Un réseau de neurones dense simple avec plusieurs couches cachées peut apprendre à partir des données de texte transformées (par exemple, des vecteurs) et peut être performant pour l'analyse de sentiments.
Un avantage majeur est que le modèle peut évoluer et améliorer ses performances avec de plus grandes quantités de données.

2. Réseau de Neurones Convolutifs (CNN)
Les CNN sont largement utilisés pour les tâches de traitement du langage naturel (NLP), bien qu'ils soient plus connus pour la vision par ordinateur. Ils sont capables d'apprendre des motifs locaux dans les données textuelles, comme des séquences de mots ou des structures syntaxiques.
Dans l'analyse de sentiments, les CNN peuvent détecter des motifs de mots ou de phrases qui sont indicatifs d'un sentiment spécifique.

3. Réseau de Neurones Récurrents (RNN) / LSTM / GRU
Pourquoi ?
Les RNN, et plus spécifiquement les LSTM (Long Short-Term Memory) et GRU (Gated Recurrent Unit), sont adaptés pour traiter des séquences de données, comme du texte.
Ces réseaux sont capables de prendre en compte le contexte et les relations à long terme entre les mots d'une critique, ce qui est essentiel pour une bonne analyse de sentiment. Par exemple, dans la phrase « Ce produit est incroyable, mais un peu cher », le contexte "incroyable" peut influencer l'analyse de sentiment, et les RNN sont bien adaptés à cela.
Les LSTM et GRU sont particulièrement populaires pour les tâches NLP car ils aident à éviter le problème de perte de mémoire à long terme des RNN classiques.

# Prétraitement des données
J'ai utilisé Tokenizer et pad_sequences pour transformer les données textuelles en vecteurs numériques adaptés aux réseaux de neurones.
Chacun des modèles utilise un Embedding pour convertir les mots en vecteurs denses, puis les passe dans les couches appropriées pour chaque architecture.

In [ ]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical

# Tokenisation des données d'entraînement et de test
tokenizer = Tokenizer(num_words=5000)
tokenizer.fit_on_texts(X_train)

# Transformation des textes en séquences numériques
X_train_seq = tokenizer.texts_to_sequences(X_train)
X_test_seq = tokenizer.texts_to_sequences(X_test)

# Padding des séquences pour qu'elles aient la même longueur
X_train_pad = pad_sequences(X_train_seq, maxlen=200)
X_test_pad = pad_sequences(X_test_seq, maxlen=200)

# Encodage des labels en catégories (pour la classification binaire)
y_train_cat = to_categorical(y_train)
y_test_cat = to_categorical(y_test)


Pour DNN, les données d'entrée doivent être aplaties (X_train_tfidf.toarray()).
Pour les autres modèles j'ai utilisé les données préparées avec pad_sequences.

J'utilise Tensorflow/Keras pour un réseau de neurones simple

# DNN

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from sklearn.metrics import classification_report, f1_score

# Création du réseau
nn = Sequential([
    Dense(128, activation='relu', input_dim=X_train_tfidf.shape[1]),
    Dropout(0.3),
    Dense(64, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compilation et entraînement
nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
nn.fit(X_train_tfidf.toarray(), y_train, epochs=10, batch_size=32, verbose=1)



/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 10s 12ms/step - accuracy: 0.8535 - loss: 0.3663
Epoch 2/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 9s 14ms/step - accuracy: 0.9272 - loss: 0.1933
Epoch 3/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 9s 12ms/step - accuracy: 0.9441 - loss: 0.1590
Epoch 4/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 10s 12ms/step - accuracy: 0.9588 - loss: 0.1217
Epoch 5/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 10s 14ms/step - accuracy: 0.9738 - loss: 0.0879
Epoch 6/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 10s 14ms/step - accuracy: 0.9821 - loss: 0.0604
Epoch 7/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 8s 12ms/step - accuracy: 0.9869 - loss: 0.0427
Epoch 8/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 11s 14ms/step - accuracy: 0.9900 - loss: 0.0341
Epoch 9/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 9s 12ms/step - accuracy: 0.9911 - loss: 0.0299
Epoch 10/10
695/695 ━━━━━━━━━━━━━━━━━━━━ 8s 12ms/step - accuracy: 0.9912 - loss: 0.0267


In [ ]:
# Évaluation
y_nn_pred = (nn.predict(X_test_tfidf.toarray()) > 0.5).astype(int)
nn_f1 = f1_score(y_test, y_nn_pred, average='weighted')
print(f"Réseau de Neurones -> F1-Score: {nn_f1:.4f}")
print("\nRapport de classification:")
print(classification_report(y_test, y_nn_pred))

174/174 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step
Réseau de Neurones -> F1-Score: 0.8777

Rapport de classification:
              precision    recall  f1-score   support

           0       0.68      0.64      0.66      1025
           1       0.92      0.93      0.93      4533

    accuracy                           0.88      5558
   macro avg       0.80      0.79      0.79      5558
weighted avg       0.88      0.88      0.88      5558



# CNN

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Conv1D, MaxPooling1D, Flatten, Dense

def build_cnn():
    model = Sequential()
    model.add(Embedding(input_dim=5000, output_dim=128, input_length=200))
    model.add(Conv1D(filters=64, kernel_size=5, activation='relu'))
    model.add(MaxPooling1D(pool_size=2))
    model.add(Flatten())
    model.add(Dense(64, activation='relu'))
    model.add(Dense(2, activation='softmax'))  # Pour une classification binaire
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

cnn_model = build_cnn()
cnn_model.summary()
cnn_model.fit(X_train_pad, y_train_cat, epochs=5, batch_size=32, validation_data=(X_test_pad, y_test_cat))

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_4 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv1d_1 (Conv1D)                    │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling1d_1 (MaxPooling1D)       │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_1 (Flatten)                  │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_13 (Dense)                     │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_14 (Dense)                     │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 48s 63ms/step - accuracy: 0.8520 - loss: 0.3625 - val_accuracy: 0.8969 - val_loss: 0.2560
Epoch 2/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 87s 70ms/step - accuracy: 0.9205 - loss: 0.2003 - val_accuracy: 0.8951 - val_loss: 0.2758
Epoch 3/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 82s 70ms/step - accuracy: 0.9540 - loss: 0.1299 - val_accuracy: 0.8816 - val_loss: 0.3180
Epoch 4/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 75s 59ms/step - accuracy: 0.9767 - loss: 0.0727 - val_accuracy: 0.8807 - val_loss: 0.4157
Epoch 5/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 90s 72ms/step - accuracy: 0.9847 - loss: 0.0480 - val_accuracy: 0.8793 - val_loss: 0.5614


In [ ]:
from sklearn.metrics import classification_report, f1_score

# Évaluation sur les données de test
def evaluate_model(model, X_test, y_test):
    y_pred = model.predict(X_test)

    # Conversion des probabilités en classes
    y_pred_classes = y_pred.argmax(axis=1)

    # Conversion des catégories one-hot en labels originaux
    y_true_classes = y_test.argmax(axis=1)

    f1 = f1_score(y_true_classes, y_pred_classes, average='weighted')
    print(f"F1-Score: {f1:.4f}")

    # Rapport complet
    print("\nRapport de classification:")
    print(classification_report(y_true_classes, y_pred_classes))

# Evaluation de CNN
evaluate_model(cnn_model, X_test_pad, y_test_cat)

174/174 ━━━━━━━━━━━━━━━━━━━━ 3s 18ms/step
F1-Score: 0.8750

Rapport de classification:
              precision    recall  f1-score   support

           0       0.71      0.59      0.64      1025
           1       0.91      0.94      0.93      4533

    accuracy                           0.88      5558
   macro avg       0.81      0.77      0.79      5558
weighted avg       0.87      0.88      0.87      5558



# RNN

In [ ]:
from tensorflow.keras.layers import SimpleRNN

def build_rnn():
    model = Sequential()
    model.add(Embedding(input_dim=5000, output_dim=128, input_length=200))
    model.add(SimpleRNN(units=64, activation='tanh'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(2, activation='softmax'))
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

rnn_model = build_rnn()
rnn_model.summary()
rnn_model.fit(X_train_pad, y_train_cat, epochs=5, batch_size=32, validation_data=(X_test_pad, y_test_cat))


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential_8"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_6 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ simple_rnn_2 (SimpleRNN)             │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_17 (Dense)                     │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_18 (Dense)                     │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 90s 118ms/step - accuracy: 0.8399 - loss: 0.3870 - val_accuracy: 0.8856 - val_loss: 0.3096
Epoch 2/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 124s 92ms/step - accuracy: 0.9122 - loss: 0.2248 - val_accuracy: 0.8766 - val_loss: 0.3166
Epoch 3/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 61s 87ms/step - accuracy: 0.9476 - loss: 0.1446 - val_accuracy: 0.8746 - val_loss: 0.3529
Epoch 4/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 81s 85ms/step - accuracy: 0.9727 - loss: 0.0853 - val_accuracy: 0.8660 - val_loss: 0.4856
Epoch 5/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 58s 84ms/step - accuracy: 0.9802 - loss: 0.0611 - val_accuracy: 0.8537 - val_loss: 0.5340


In [ ]:
# Evaluation de RNN
evaluate_model(rnn_model, X_test_pad, y_test_cat)

174/174 ━━━━━━━━━━━━━━━━━━━━ 3s 18ms/step
F1-Score: 0.8500

Rapport de classification:
              precision    recall  f1-score   support

           0       0.62      0.54      0.58      1025
           1       0.90      0.92      0.91      4533

    accuracy                           0.85      5558
   macro avg       0.76      0.73      0.74      5558
weighted avg       0.85      0.85      0.85      5558



# LSTM

In [ ]:
from tensorflow.keras.layers import LSTM

def build_lstm():
    model = Sequential()
    model.add(Embedding(input_dim=5000, output_dim=128, input_length=200))
    model.add(LSTM(units=64))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(2, activation='softmax'))
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

lstm_model = build_lstm()
lstm_model.summary()
lstm_model.fit(X_train_pad, y_train_cat, epochs=5, batch_size=32, validation_data=(X_test_pad, y_test_cat))


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential_9"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_7 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lstm_3 (LSTM)                        │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_19 (Dense)                     │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_20 (Dense)                     │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 118s 166ms/step - accuracy: 0.8542 - loss: 0.3570 - val_accuracy: 0.8974 - val_loss: 0.2761
Epoch 2/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 162s 195ms/step - accuracy: 0.9165 - loss: 0.2147 - val_accuracy: 0.8976 - val_loss: 0.2595
Epoch 3/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 141s 203ms/step - accuracy: 0.9352 - loss: 0.1715 - val_accuracy: 0.8946 - val_loss: 0.2842
Epoch 4/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 114s 163ms/step - accuracy: 0.9462 - loss: 0.1448 - val_accuracy: 0.8874 - val_loss: 0.3182
Epoch 5/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 111s 160ms/step - accuracy: 0.9584 - loss: 0.1183 - val_accuracy: 0.8850 - val_loss: 0.3775


In [ ]:
# Evaluation de LSTM
evaluate_model(lstm_model, X_test_pad, y_test_cat)


174/174 ━━━━━━━━━━━━━━━━━━━━ 8s 47ms/step
F1-Score: 0.8796

Rapport de classification:
              precision    recall  f1-score   support

           0       0.74      0.58      0.65      1025
           1       0.91      0.95      0.93      4533

    accuracy                           0.89      5558
   macro avg       0.82      0.77      0.79      5558
weighted avg       0.88      0.89      0.88      5558



# GRU

In [ ]:
from tensorflow.keras.layers import GRU

def build_gru():
    model = Sequential()
    model.add(Embedding(input_dim=5000, output_dim=128, input_length=200))
    model.add(GRU(units=64))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(2, activation='softmax'))
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

gru_model = build_gru()
gru_model.summary()
gru_model.fit(X_train_pad, y_train_cat, epochs=5, batch_size=32, validation_data=(X_test_pad, y_test_cat))


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential_11"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_8 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ gru (GRU)                            │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_24 (Dense)                     │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_25 (Dense)                     │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 116s 163ms/step - accuracy: 0.8459 - loss: 0.3711 - val_accuracy: 0.8951 - val_loss: 0.2676
Epoch 2/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 138s 157ms/step - accuracy: 0.9184 - loss: 0.2097 - val_accuracy: 0.8958 - val_loss: 0.2587
Epoch 3/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 142s 158ms/step - accuracy: 0.9359 - loss: 0.1734 - val_accuracy: 0.8913 - val_loss: 0.2873
Epoch 4/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 143s 159ms/step - accuracy: 0.9495 - loss: 0.1284 - val_accuracy: 0.8890 - val_loss: 0.3202
Epoch 5/5
695/695 ━━━━━━━━━━━━━━━━━━━━ 146s 166ms/step - accuracy: 0.9596 - loss: 0.1041 - val_accuracy: 0.8802 - val_loss: 0.3905


In [ ]:
# Evaluation de GRU
evaluate_model(gru_model, X_test_pad, y_test_cat)

174/174 ━━━━━━━━━━━━━━━━━━━━ 13s 66ms/step
F1-Score: 0.8735

Rapport de classification:
              precision    recall  f1-score   support

           0       0.73      0.55      0.63      1025
           1       0.90      0.95      0.93      4533

    accuracy                           0.88      5558
   macro avg       0.82      0.75      0.78      5558
weighted avg       0.87      0.88      0.87      5558



# Modèle Performant

Meilleur modèle : SVM

- Performance globale : Le SVM a la meilleure Accuracy (0.8967) et le meilleur F1-Score (0.8915).
Ces métriques indiquent que SVM gère bien les déséquilibres de classes et offre une précision élevée.

- Stabilité : SVM est robuste et souvent performant sur des ensembles de données textuelles ou transformées via TF-IDF.
Il est moins sujet au surapprentissage que les modèles complexes comme DNN, CNN, ou LSTM lorsque le volume de données est limité.

- Simplicité : Comparé aux réseaux neuronaux, le SVM est plus facile à configurer et interpréter.
Il nécessite moins de puissance de calcul et est rapide à entraîner.




# Alternatives :


Régression logistique : Très proche de SVM en termes de performance.
On peut l'utiliser si on privilégie un modèle plus simple avec des performances légèrement inférieures.

Pour les réseaux de neurones :
LSTM : Si On veut un modèle basé sur des réseaux neuronaux, LSTM est le plus performant parmi les modèles séquentiels (F1-Score : 0.8796).

In [51]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import SVC
import joblib

# Initialiser le vectoriseur TF-IDF
tfidf = TfidfVectorizer(max_features=12695)  # Utilisez les mêmes paramètres que lors de l'entraînement initial

# Appliquer la vectorisation sur les données d'entraînement
X_train_tfidf = tfidf.fit_transform(X_train)

# Entraîner le modèle SVM
svm_model = SVC()
svm_model.fit(X_train_tfidf, y_train)

# Sauvegarder le modèle et le vectoriseur
joblib.dump(svm_model, "svm_model.pkl")
joblib.dump(tfidf, "vectorizer.pkl")

['vectorizer.pkl']

In [54]:
import streamlit as st
import joblib

code = """
import streamlit as st
import joblib

# Charger le vectoriseur et le modèle SVM sauvegardés
vectorizer = joblib.load("vectorizer.pkl")  # Charger le même TF-IDF vectorizer utilisé lors de l'entraînement
svm_model = joblib.load("svm_model.pkl")   # Charger le modèle SVM

# Fonction pour effectuer une prédiction
def predict_sentiment(user_input):
    text_features = vectorizer.transform([user_input])  # Transformer l'entrée utilisateur
    prediction = svm_model.predict(text_features)       # Faire une prédiction
    return prediction[0]

# Interface utilisateur
st.title("Analyse de Sentiment")
st.write("Entrez un avis pour prédire le sentiment.")

# Entrée utilisateur
user_input = st.text_input("Votre avis :", "")

# Bouton pour prédire
if st.button("Prédire"):
    if user_input.strip():
        result = predict_sentiment(user_input)
        st.write(f"Sentiment prédit : **{result}**")
    else:
        st.write("Veuillez entrer un avis.")
"""

# Enregistrer le code dans un fichier app.py
with open("app.py", "w") as file:
    file.write(code)
print("Fichier app.py créé avec succès.")


# Charger le vectoriseur et le modèle SVM sauvegardés
vectorizer = joblib.load("vectorizer.pkl")  # Charger le même TF-IDF vectorizer utilisé lors de l'entraînement
svm_model = joblib.load("svm_model.pkl")   # Charger le modèle SVM

# Fonction pour effectuer une prédiction
def predict_sentiment(user_input):
    text_features = vectorizer.transform([user_input])  # Transformer l'entrée utilisateur
    prediction = svm_model.predict(text_features)       # Faire une prédiction
    return prediction[0]

# Interface utilisateur
st.title("Analyse de Sentiment")
st.write("Entrez un avis pour prédire le sentiment.")

# Entrée utilisateur
user_input = st.text_input("Votre avis :", "")

# Bouton pour prédire
if st.button("Prédire"):
    if user_input.strip():
        result = predict_sentiment(user_input)
        st.write(f"Sentiment prédit : **{result}**")
    else:
        st.write("Veuillez entrer un avis.")

2024-12-16 14:46:51.538 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-16 14:46:51.539 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-16 14:46:51.542 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-16 14:46:51.545 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-16 14:46:51.547 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-16 14:46:51.548 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-16 14:46:51.550 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-16 14:46:51.551 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bar

Fichier app.py créé avec succès.
